core/matrix/ell.cpp

```diff
@@ -0,0 +1,126 @@
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include "core/matrix/ell.hpp"
+
+
+#include <algorithm>
+
+
+#include "core/base/exception_helpers.hpp"
+#include "core/base/executor.hpp"
+#include "core/base/math.hpp"
+#include "core/base/utils.hpp"
+#include "core/matrix/dense.hpp"
+
+
+namespace gko {
+namespace matrix {
+
+
+template <typename ValueType, typename IndexType>
+void Ell<ValueType, IndexType>::apply(const LinOp *b, LinOp *x) const
+    NOT_IMPLEMENTED;
+
+
+template <typename ValueType, typename IndexType>
+void Ell<ValueType, IndexType>::apply(const LinOp *alpha, const LinOp *b,
+                                      const LinOp *beta, LinOp *x) const
+    NOT_IMPLEMENTED;
+
+
+template <typename ValueType, typename IndexType>
+void Ell<ValueType, IndexType>::convert_to(Dense<ValueType> *other) const
+    NOT_IMPLEMENTED;
+
+
+template <typename ValueType, typename IndexType>
+void Ell<ValueType, IndexType>::move_to(Dense<ValueType> *other)
+    NOT_IMPLEMENTED;
+
+
+template <typename ValueType, typename IndexType>
+void Ell<ValueType, IndexType>::read_from_mtx(const std::string &filename)
+{
+    auto data = read_raw_from_mtx<ValueType, IndexType>(filename);
+
+    // Get the maximum number of nonzero elements of every row.
+    size_type nnz = 0;
+    index_type current_row = 0;
+    size_type max_nonzeros_per_row = 0;
+    for (const auto &elem : data.nonzeros) {
+        if (std::get<0>(elem) != current_row) {
+            max_nonzeros_per_row = std::max(max_nonzeros_per_row, nnz);
+            nnz = 0;
+        }
+        nnz += (std::get<2>(elem) != zero<ValueType>());
+    }
+
+    // Create an ELLPACK format matrix based on the sizes.
+    auto tmp = create(this->get_executor()->get_master(), data.num_rows,
+                      data.num_cols, max_nonzeros_per_row, data.num_rows);
+
+    // Get values and column indexes.
+    size_type ind = 0;
+    int n = data.nonzeros.size();
+    auto vals = tmp->get_values();
+    auto col_idxs = tmp->get_col_idxs();
+    for (size_type row = 0; row < data.num_rows; row++) {
+        size_type col = 0;
+        while (ind < n && std::get<0>(data.nonzeros[ind]) == row) {
+            auto val = std::get<2>(data.nonzeros[ind]);
+            if (val != zero<ValueType>()) {
+                tmp->val_at(row, col) = val;
+                tmp->col_at(row, col) = std::get<1>(data.nonzeros[ind]);
+                col++;
+            }
+            ind++;
+        }
+        for (auto i = col; i < max_nonzeros_per_row; i++) {
+            tmp->val_at(row, i) = zero<ValueType>();
+            tmp->col_at(row, i) = (i == 0) ? 0 : tmp->col_at(row, i-1);
```

Comment:
@hartwiganzt I didn't get a response for my last question here. Could you comment on this:

> I think the standard for ELL is to set all the unused values to 0, @hartwiganzt what do you think?

I was of course referring to the values of column indexes.

Thanks!

Reply:
I just had a discussion with @gflegar about this: The standard is not very specific about this, Your attempt is to use the last index to probably get better performance (the corresponding x-value is already in cache). At the same time, putting in zeros makes it easier to comprehend. Obviously, the values have to be zero too. Let's go that way!

Reply:
Sorry, @hartwiganzt which way exactly are we using?

Reply:
The padding entries have
val: 0 0 0 ...
colidx: 0 0 0 ...

Reply:
fixed